File: 2022/aoc_python/day_5.py
import copy
import re

from lib import io

print("Advent of Code 2022 Day 5")
print("-------------------------")
file_contents = io.file_to_list("input/day-5-input.txt")

# Figure out where initial conditions end and instructions begin.
i = 0
while file_contents[i]:
    i += 1

# Split input into initial conditions and instructions.
initial_conditions = file_contents[:i]  # Strips out the empty line.
instructions = file_contents[i + 1:]    # See above.

# Construct stacks and their initial contents.
max_length = len(initial_conditions[0])         # Guaranteed all lines in section are same length.
stack_labels = initial_conditions[-1].split()   # At bottom of section.
# print(f"stack_labels: {stack_labels}")
stacks_for_single_move_crane = { k: [] for k in stack_labels}

stack_contents = initial_conditions[:-1]
stack_pattern = re.compile(r"\[\w\]")
for level in reversed(stack_contents):
    # Sweep a window of size three across level; this captures the crate if present.
    cur_level_index = 0
    cur_stack_index = 0
    while cur_level_index <= len(level) - 3:
        cur_stack_contents = level[cur_level_index:cur_level_index + 3]
        # print(f"cur_stack_contents: {cur_stack_contents}")
        if stack_pattern.match(cur_stack_contents):
            # Crate found; push it onto the current stack.
            # print("Crate found.\n")
            stacks_for_single_move_crane[stack_labels[cur_stack_index]].append(level[cur_level_index + 1])
        # else:
            # No crate found; advance to next stack.
            # print("Crate not found.\n")

        cur_stack_index += 1
        cur_level_index += 4    # Skip over the space separating each stack.

stacks_for_multi_move_crane = copy.deepcopy(stacks_for_single_move_crane)
# print(f"stacks_for_single_move_crane (initial): {stacks_for_single_move_crane}")

# Now parse the instructions.
for instruction in instructions:
    # print(f"instruction: {instruction}")
    instruction_list = instruction.split()  # All instructions will have 6 elements when split.
    num_crates = int(instruction_list[1])
    source_stack = instruction_list[3]
    target_stack = instruction_list[5]

    # Move crates according to instructions.
    for i in range(num_crates):
        crate = stacks_for_single_move_crane[source_stack].pop()
        stacks_for_single_move_crane[target_stack].append(crate)

# print(f"stacks_for_single_move_crane (final): {stacks_for_single_move_crane}")

# Get list of crates on top.
top_crates = ""
for stack in stacks_for_single_move_crane.values():
    top_crates += stack.pop()

print("PART 1")
print("======")
print(f"Crates on top: {top_crates}")
print("======")
# Attempt 1: TWSGQHNHL (correct)

# print(f"stacks_for_multi_move_crane (initial): {stacks_for_multi_move_crane}")
# Parse the instructions assuming crane preserves order of removed crates.
for instruction in instructions:
    # print(f"instruction: {instruction}")
    instruction_list = instruction.split()  # All instructions will have 6 elements when split.
    num_crates = int(instruction_list[1])
    source_stack = instruction_list[3]
    target_stack = instruction_list[5]

    # Move crates according to instructions.
    # Moving multiple crates whilst preserving order can be modelled by pushing the crates onto
    # a "buffer stack" and then emptying the buffer stack onto the target stack.
    buffer_stack = []
    for i in range(num_crates):
        buffer_stack.append(stacks_for_multi_move_crane[source_stack].pop())

    while buffer_stack:
        stacks_for_multi_move_crane[target_stack].append(buffer_stack.pop())

# print(f"stacks_for_multi_move_crane (final): {stacks_for_multi_move_crane}")

# Get list of crates on top.
top_crates = ""
for stack in stacks_for_multi_move_crane.values():
    top_crates += stack.pop()

print("PART 2")
print("======")
print(f"Crates on top: {top_crates}")

File: 2022/aoc_scala/Day5.scala
import scala.io.Source
import scala.collection.mutable.{HashMap, Stack, StringBuilder}

def copyStackMap(stackMap: HashMap[String, Stack[String]]) =
    // Implements a deep copy function for the stack maps.
    val stackMapCopy = HashMap.empty[String, Stack[String]]
    for (label, stack) <- stackMap
    do
        // Use a stack to preserve order of items in stack when transferring them.
        val buffer = Stack[String]()
        while (!stack.isEmpty) {
            buffer.push(stack.pop())
        }
        while (!buffer.isEmpty) {
            val itemToCopy = buffer.pop()
            stack.push(itemToCopy)
            stackMapCopy.getOrElseUpdate(label, Stack[String]()).push(itemToCopy)
        }

    stackMapCopy

def getTopCrates(stackMap: HashMap[String, Stack[String]]) =
    // Returns a string representation of all crates at the top of the stacks.
    val topCrates = new StringBuilder()
    for stack <- stackMap.values
    do
        topCrates.append(stack.pop())

    topCrates.result()

@main def dayFive() =
    println("Advent of Code 2022 Day 5")
    println("-------------------------")
    val fileContents = Source.fromFile("input/day-5-input.txt").getLines().toList
    val dividerIndex = fileContents.indexOf("") // Where initial conditions end and instructions begin.
    var (initialConditions, instructions) = fileContents.splitAt(dividerIndex)
    instructions = instructions.tail    // Strip off empty element.

    // Construct stacks and their initial contents.
    val maxLength = initialConditions.head.length()                 // Guaranteed all lines in section are same length.
    val stackLabels = initialConditions.last.split(" ")
                                            .map(_.trim())
                                            .filter(_.nonEmpty)   // At bottom of section.
    val stackContents = initialConditions.init
    val stackPattern = """\[(\w)\]""".r

    val stacksForSingleMoveCrane = HashMap.empty[String, Stack[String]]
    stackLabels.foreach{ label =>
            stacksForSingleMoveCrane += (label -> Stack[String]())
    }

    for level <- stackContents.reverse
    do
        // Sweep a window of size three across level; this captures the crate if present.
        var curLevelIndex = 0
        var curStackIndex = 0
        while (curLevelIndex <= level.length() - 3) {
            val curStackContents = level.substring(curLevelIndex, curLevelIndex + 3)
            curStackContents match
                case stackPattern(crateLabel) =>    // Crate found; push it onto the current stack.
                    stacksForSingleMoveCrane(stackLabels(curStackIndex)).push(crateLabel)
                case _ => ()                        // No crate found; advance to next stack.

            curStackIndex += 1
            curLevelIndex += 4  // Skip over the space separating each stack.
        }

    val stacksForMultiMoveCrane = copyStackMap(stacksForSingleMoveCrane)

    // Now parse the instructions.
    for instruction <- instructions
    do
        val instructionList = instruction.split(" ")    // All instructions will have 6 elements when split.
        val numCrates = instructionList(1).toInt
        val sourceStack = instructionList(3)
        val targetStack = instructionList(5)

        // Move crates according to instructions.
        for _ <- 0 until numCrates
        do
            stacksForSingleMoveCrane(targetStack).push(
                stacksForSingleMoveCrane(sourceStack).pop()
            )

    println("PART 1")
    println("======")
    println(f"Crates on top: ${getTopCrates(stacksForSingleMoveCrane)}")
    println("======")

    // Parse the instructions assuming crane preserves order of removed crates.
    for instruction <- instructions
    do
        val instructionList = instruction.split(" ")    // All instructions will have 6 elements when split.
        val numCrates = instructionList(1).toInt
        val sourceStack = instructionList(3)
        val targetStack = instructionList(5)

        // Move crates according to instructions.
        // Moving multiple crates whilst preserving order can be modelled by pushing the crates onto
        // a "buffer stack" and then emptying the buffer stack onto the target stack.
        val bufferStack = Stack[String]()
        for _ <- 0 until numCrates
        do
            bufferStack.push(
                stacksForMultiMoveCrane(sourceStack).pop()
            )

        while (!bufferStack.isEmpty) {
            stacksForMultiMoveCrane(targetStack).push(
                bufferStack.pop()
            )
        }

    println("PART 2")
    println("======")
    println(f"Crates on top: ${getTopCrates(stacksForMultiMoveCrane)}")
